### PR TITLE
fix(datastore): Switch to BFS in Join Builder for Alias, fixes - #2488

### PR DESCRIPTION
**Pull request by ankpshah**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*
#2488 
*Description of changes:*
Replace DFS Join Builder with BFS to Improve Alias Management in Joins

- Replace the recursiveBuildJoins method (DFS-based) with buildJoinsUsingBFS to address issues with incorrect table alias usage in WHERE conditions.
- The BFS approach ensures that all tables directly related to the root table are assigned their original names without any numeric suffixes (alias), enhancing the readability and correctness of SQL queries, especially in WHERE clauses involving intermediate tables in multilevel joins.
- Implement a tracking mechanism for visited (table, FK) combinations to support multiple foreign keys and prevent infinite loops in cyclic relationships.
- This change aims to improve the reliability of SQL query construction, particularly by ensuring that WHERE conditions on intermediate tables in multilevel joins are more intuitive and less error-prone.

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [X] Added Unit Tests
- [ ] Added Integration Tests
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tylerjroach** [comment]
Changes Look great. Just as a reminder, we discussed addition of query using CPK, as well as multiple statements in where clause.

**Comment by ankpshah** [comment]
Thanks. I have added tests for querying using CPK, as well as multiple conditions in where statement.

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java [hunk 1]
```diff
@@ -225,6 +243,299 @@ public void querySavedDataWithMultipleForeignKeysOfSameType() throws DataStoreEx
...
+    }
+
+    /**
+     * Test that querying the saved item with multiple paths for foreign keys to the same
+     * model along with Multiple conditions and Custom Primary Key correctly forms the where clause
+     * and returns expected data.
+     * @throws DataStoreException On unexpected failure manipulating items in/out of DataStore.
+     */
+    @Test
+    public void querySavedDataWithMultipleJoinPathsWithMultipleConditionsAndCPK() throws DataStoreException {
```
Comment (tylerjroach) [inline aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java:418]: Can you adjust this query to use the composite key: https://docs.amplify.aws/android/build-a-backend/more-features/datastore/customize-primary-keys/#querying-records-with-custom-primary-keys

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java [hunk 2]
```diff
@@ -225,6 +243,299 @@ public void querySavedDataWithMultipleForeignKeysOfSameType() throws DataStoreEx
...
+
+
+    /**
+     * Test that querying the saved item with multiple paths for foreign keys to the same
+     * model with missing data for intermediate model in join path and Custom Primary Key
+     * correctly forms the where clause and returns expected data.
+     * @throws DataStoreException On unexpected failure manipulating items in/out of DataStore
+     */
+    @Test
+    public void queryMultipleJoinWithMultipleRecordsWithMultipleConditionsAndCPK()
```
Comment (tylerjroach) [inline aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java:461]: Can you adjust this query to use the composite key: https://docs.amplify.aws/android/build-a-backend/more-features/datastore/customize-primary-keys/#querying-records-with-custom-primary-keys

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java [hunk 3]
```diff
@@ -265,6 +576,7 @@ public void querySavedDataWithMultiLevelJoins() throws DataStoreException {
         assertEquals(comments.get(0).getPost().getBlog().getOwner(), blogOwner);
     }
 
+
```
Comment (tylerjroach) [inline aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java:579]: extra line

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java [hunk 4]
```diff
@@ -225,6 +243,299 @@ public void querySavedDataWithMultipleForeignKeysOfSameType() throws DataStoreEx
...
+                        .and(PostCPK.TITLE.eq("Beta Post")))));
+        assertEquals(1, posts.size());
+        assertTrue(posts.contains(postByBeta));
+
+        posts = adapter.query(PostCPK.class,
+                Where.matches(PostCPK.BLOG_OWNER.eq(expectedGammaBlogOwnerCPK).and(PostCPK.RATING.gt(4)
+                        .and(PostCPK.TITLE.eq("Gamma Post")))));
+        assertEquals(1, posts.size());
+        assertTrue(posts.contains(postByGamma));
+
```
Comment (tylerjroach) [inline aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java:536]: extra line

**Review by tylerjroach — CHANGES_REQUESTED** [review]

File: aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteCommandFactory.java [hunk 5]
```diff
@@ -656,4 +597,140 @@ private boolean shouldCreateIndex(ModelIndex modelIndex, Map<String, ModelAssoci
...
+                                    String rootTableAlias) {
+        Queue<TableInfo> queue = new LinkedList<>();
+        queue.add(new TableInfo(rootTable, rootTableAlias));
+
+        // Use a Set to track visited (table, foreign key) combinations to allow multiple references.
+        Set<String> visitedCombinations = new HashSet<>();
+
+        while (!queue.isEmpty()) {
+            TableInfo currentInfo = queue.poll();
+            if (currentInfo == null) {
```
Comment (mattcreaser) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteCommandFactory.java:644]: This condition isn't possible, can be removed.

**Review by mattcreaser — COMMENTED** [review]
Just a few minor comments to add.

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java (hunk 1 above)
Comment (ankpshah) [reply aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterQueryTest.java]: I have added Composite Key for Post in Query.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteCommandFactory.java [hunk 6]
```diff
@@ -656,4 +597,140 @@ private boolean shouldCreateIndex(ModelIndex modelIndex, Map<String, ModelAssoci
...
+
+        public SQLiteTable getSQLiteTable() {
+            if (this.table == null) {
+                throw new IllegalStateException("SQLiteTable has not been initialized.");
+            }
+            return table;
+        }
+
+        public String getAlias() {
+            return alias;
```
[the remark this answers is not in the document]
Comment (ankpshah) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteCommandFactory.java:727]: Yes you are right. Alias isn't expected to be null. I have added a check for it in constructor. Thanks!

**Review by mattcreaser — APPROVED** [review]
🎉

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**